ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java

```diff
@@ -59,15 +59,30 @@ public BonsaiWorldStateArchive(
       final StorageProvider provider, final Blockchain blockchain, final long maxLayersToLoad) {
     this.blockchain = blockchain;
 
-    worldStateStorage = new BonsaiWorldStateKeyValueStorage(provider);
-    persistedState = new BonsaiPersistedWorldState(this, worldStateStorage);
-    layeredWorldStatesByHash = new HashMap<>();
+    this.worldStateStorage = new BonsaiWorldStateKeyValueStorage(provider);
+    this.persistedState = new BonsaiPersistedWorldState(this, worldStateStorage);
+    this.layeredWorldStatesByHash = new HashMap<>();
     this.maxLayersToLoad = maxLayersToLoad;
+    blockchain.observeBlockAdded(
+        event -> {
+          if (event.isNewCanonicalHead()) {
```

Comment:
(optional) Maybe an `if(!event.isCanonicalHead()) return;` here? Makes it easier to scan.